### PR TITLE
Ddf 1723 Update platform-parser-api to support all marshal/unmarshal paths

### PR DESCRIPTION
**Pull request by wbarnie**
@codice/green

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/codice/ddf/pull/395" target="_blank">ReviewNinja

File: platform/parser/xml/src/test/java/org/codice/ddf/parser/xml/TestXmlParser.java [hunk 1]
```diff
@@ -38,13 +23,34 @@
...
+import static org.custommonkey.xmlunit.XMLAssert.*;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.not;
+import static org.junit.Assert.assertThat;
 
 public class TestXmlParser {
     private Parser parser;
 
     private ParserConfigurator configurator;
+   // private ParserConfigurator badConfigurator;
```
Comment (roelens8) [inline platform/parser/xml/src/test/java/org/codice/ddf/parser/xml/TestXmlParser.java]: Remove commented code.

File: platform/security/pdp/security-pdp-xacmlrealm/pom.xml [hunk 2]
```diff
@@ -25,6 +27,16 @@
 
     <dependencies>
```
Comment (coyotesqrl) [inline platform/security/pdp/security-pdp-xacmlrealm/pom.xml]: I won't have time to properly review this until tomorrow or Monday, but this class shouldn't be included in this PR.

File: platform/parser/xml/src/test/java/org/codice/ddf/parser/xml/TestXmlParser.java [hunk 3]
```diff
@@ -59,9 +69,13 @@
     public void setUp() throws Exception {
         parser = new XmlParser();
 
-        List<String> ctxPath = ImmutableList.of(MotherElement.class.getPackage().getName());
+        List<String> ctxPath = ImmutableList.of(MotherElement.class.getPackage()
+                .getName());
         configurator = parser.configureParser(ctxPath, TestXmlParser.class.getClassLoader());
 
+       /* badConfigurator = parser.configureParser(ctxPath, TestXmlParser.class.getClassLoader());
```
Comment (roelens8) [inline platform/parser/xml/src/test/java/org/codice/ddf/parser/xml/TestXmlParser.java]: Remove commented code if not needed.

**Comment by coyotesqrl** [comment]
@kcwire @shane-voisard Could you two give this PR a glance please? I've already seen and touched quite a bit of the new Parser code so I'm worried I might miss something.

**Comment by wbarnie** [comment]
@kcwire @shane-voisard updated the code

File: platform/parser/api/src/main/java/org/codice/ddf/parser/Parser.java [hunk 4]
```diff
@@ -41,24 +46,52 @@
...
-     * @param obj the root of the object graph to convert
-     * @param os the output stream on which the converted object is written
+     * @param obj          the root of the object graph to convert
+     * @param os           the output stream on which the converted object is written
      * @throws ParserException
      */
     void marshal(ParserConfigurator configurator, Object obj, OutputStream os)
             throws ParserException;
 
     /**
```
Comment (oscaritoro) [inline platform/parser/api/src/main/java/org/codice/ddf/parser/Parser.java]: You probably want to add a description of what this marshal does... something like `Converts an object graph into the appropriate output format, writing it to the given node.`

File: platform/parser/api/src/main/java/org/codice/ddf/parser/Parser.java [hunk 5]
```diff
@@ -41,24 +46,52 @@
...
+     * @param cls          {@code Class} for the conversion
+     * @param stream       input stream which is read for the object data
+     * @param <T>          expected return object
      * @return an object of type {@code T} as read and converted from the stream
      * @throws ParserException
      */
     <T> T unmarshal(ParserConfigurator configurator, Class<? extends T> cls, InputStream stream)
             throws ParserException;
+
+    /**
```
Comment (oscaritoro) [inline platform/parser/api/src/main/java/org/codice/ddf/parser/Parser.java]: You probably want to also update the javadocs of these two new unmarshal methods as well...

File: platform/parser/api/src/main/java/org/codice/ddf/parser/Parser.java [hunk 6]
```diff
@@ -41,24 +46,52 @@
...
+     * @param <T>
+     * @param cls
+     * @param node
+     * @return an object of type {@code T} as read and converted from the stream
+     * @throws ParserException
+     */
+    <T> JAXBElement<T> unmarshal(ParserConfigurator configurator, Class<? extends T> cls, Node node)
```
Comment (oscaritoro) [inline platform/parser/api/src/main/java/org/codice/ddf/parser/Parser.java]: How come the unmarshal for nodes and sources returns a JAXBElement<T>, but the unmarshal for streams a "T". Is this on purpose?

**Comment by kcwire** [comment]
It appears you are not using the correct formatter. Please update the PR using the correct formatter.

File: platform/parser/api/src/main/java/org/codice/ddf/parser/Parser.java (hunk 6 above)
Comment (coyotesqrl) [inline platform/parser/api/src/main/java/org/codice/ddf/parser/Parser.java]: I'm assuming that the callers need the `JAXBElement`s for their own purposes. If not, then this should be changed to return `T`. It isn't terrible if it stays like this _if it needs to be like this_ as this has become tightly bound to XML/JAXB at this point.

File: platform/parser/xml/src/main/java/org/codice/ddf/parser/xml/XmlParser.java [hunk 7]
```diff
@@ -69,20 +76,89 @@ public JAXBContext load(CacheKey cacheKey) throws Exception {
...
+                return unmarshal;
+            } catch (XMLStreamException | JAXBException e) {
+                throw new RuntimeException("Error unmarshalling", e);
+            }
+        });
+    }
+
+    @Override
+    public <T> T unmarshal(ParserConfigurator configurator, Class<? extends T> cls,
+            Node node) throws ParserException {
```
Comment (oscaritoro) [inline platform/parser/xml/src/main/java/org/codice/ddf/parser/xml/XmlParser.java]: Just noticed that the InputStream version of unmarshal declares the third parameter as "final", should the Node and Source versions do this as well?

Comment (wbarnie) [inline platform/parser/xml/src/main/java/org/codice/ddf/parser/xml/XmlParser.java]: fixed

File: platform/parser/api/src/main/java/org/codice/ddf/parser/Parser.java (hunk 6 above)
Comment (wbarnie) [inline platform/parser/api/src/main/java/org/codice/ddf/parser/Parser.java]: removed use of JAXBElement

**Comment by wbarnie** [comment]
@kcwire : I did not have the correct formatter at first but I made sure to use the correct formatter on the latest commits. It just looks weird in a couple of long lines.

**Comment by oscaritoro** [comment]
LGTM :+1:

File: platform/parser/xml/src/main/java/org/codice/ddf/parser/xml/XmlParser.java [hunk 8]
```diff
@@ -69,20 +76,89 @@ public JAXBContext load(CacheKey cacheKey) throws Exception {
...
+                @SuppressWarnings("unchecked")
+                T unmarshal = (T) unmarshaller.unmarshal(source);
+                return unmarshal;
+            } catch (JAXBException e) {
+                throw new RuntimeException("Error unmarshalling", e);
+            }
+        });
+    }
+
+    private void marshal(ParserConfigurator configurator, Consumer<Marshaller> callable)
```
Comment (coyotesqrl) [inline platform/parser/xml/src/main/java/org/codice/ddf/parser/xml/XmlParser.java]: Should rename this parameter to something like `marshalConsumer` or `consumer`.

Comment (wbarnie) [inline platform/parser/xml/src/main/java/org/codice/ddf/parser/xml/XmlParser.java]: Renamed callable to marshallerConsumer

**Comment by wbarnie** [comment]
Richard i re-based and squashed

**Comment by coyotesqrl** [comment]
:+1:

**Comment by roelens8** [comment]
:+1:

**Comment by jckilmer** [comment]
Heroed: DDF successfully built and deployed.

**Comment by wbarnie** [comment]
@kcwire @shane-voisard I need one more person to review, then merge the code.  Richard said that you had the ability to review and merge.  Could you please do this for me?

File: platform/parser/api/src/main/java/org/codice/ddf/parser/Parser.java (hunk 4 above)
Comment (oscaritoro) [inline platform/parser/api/src/main/java/org/codice/ddf/parser/Parser.java]: !fixed

File: platform/parser/api/src/main/java/org/codice/ddf/parser/Parser.java (hunk 5 above)
Comment (oscaritoro) [inline platform/parser/api/src/main/java/org/codice/ddf/parser/Parser.java]: !fixed

File: platform/parser/api/src/main/java/org/codice/ddf/parser/Parser.java (hunk 6 above)
Comment (oscaritoro) [inline platform/parser/api/src/main/java/org/codice/ddf/parser/Parser.java]: !fixed
